pkg/reconciler/build/resources/pod.go

```diff
@@ -491,3 +525,397 @@ func getFailureMessage(pod *corev1.Pod) string {
+	}
+
+	if serviceAccountName == "" || serviceAccountName == "default" {
+		// GKE metadata server authentication
+		tokens, err := getGCRAuthorizationKey()
+		if err != nil {
+			return nil, err
+		}
+
+		imgopt = remote.WithAuth(
+			&authn.Bearer{Token: tokens[0].AccessToken})
+
+	} else {
+		// TODO(aaron-prindle) make sure to try all imagePullSecrets/registries
+		// TODO(aaron-prindle) see if there is a better way than blocking
+		WaitForServiceAccount(kubeclient, serviceAccountName, build.Namespace, "desc")
+		sa, err := kubeclient.CoreV1().ServiceAccounts(build.Namespace).Get(serviceAccountName, metav1.GetOptions{})
+		if err != nil {
+			return nil, err // TODO(aaron-prindle) better err msg? wrap errors?
+		}
+
+		for _, secret := range sa.ImagePullSecrets {
+			// TODO(aaron-prindle) see if there is a better way than blocking
+			WaitForSecret(kubeclient, secret.Name, build.Namespace, "desc")
+			scrt, err := kubeclient.CoreV1().Secrets(build.Namespace).Get(secret.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err // TODO(aaron-prindle) better err msg? wrap errors?
+			}
+
+			// TODO(aaron-prindle) check scrt.Type directly for type checking?
+			if _, ok := scrt.Data[".dockerconfigjson"]; ok {
+				dockerconfigjson := scrt.Data[".dockerconfigjson"]
+				var dat dockerJSON
+				if err := json.Unmarshal(dockerconfigjson, &dat); err != nil {
+					return nil, err
+				}
+				for _, authEnt := range dat.Auths {
+					decodedauth, err := base64.StdEncoding.DecodeString(authEnt.Auth)
+					if err != nil {
+						return nil, err
+					}
+					decodedauthstr := string(decodedauth)
+					decodedauthstrcut := decodedauthstr[10:]
+
+					// convert to yaml and strip beginning part `_json_key:`
+					var dockerscrt dockerSecret
+					if err := json.Unmarshal([]byte(decodedauthstrcut), &dockerscrt); err != nil {
+						return nil, err
+					}
+
+					// dockerscrt.PrivateKey
+					imgopt = remote.WithAuth(
+						&authn.Basic{
+							Username: "_json_key",
+							Password: decodedauthstrcut,
+						})
+				}
+			} else if _, ok := scrt.Data[".dockercfg"]; ok {
+				dockercfg := scrt.Data[".dockercfg"]
+				var dat dockerCfg
+				if err := json.Unmarshal(dockercfg, &dat); err != nil {
+					return nil, err
+				}
+				// for registry, cfgEnt := range dat {
+				for _, cfgEnt := range dat {
+					imgopt = remote.WithAuth(
+						&authn.Basic{
+							Username: cfgEnt.Username,
+							Password: cfgEnt.Password,
+						})
+				}
+			} else {
+				return nil, fmt.Errorf("no usable secret values was found for image %s", image)
+			}
+		}
+	}
+
+	img, err = remote.Image(ref, imgopt)
+	if err != nil {
+		return nil, fmt.Errorf("error with remote.Image %v", err)
+	}
+	cfg, err := img.ConfigFile()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get config for image %s: %v", image, err)
+	}
+	cache.set(image, cfg.Config.Entrypoint)
+	return cfg.Config.Entrypoint, nil
```

Comment:
i think this function needs to get broken up into a few functions (with unit test coverage)